File: #201_Bitwise_AND_of_Numbers_Range/Solution.java
public class Solution {
    public int rangeBitwiseAnd(int m, int n) {
        if (m == n) return m;
        if (m == 0) return 0;
        int shift = 0;
        while (m != n) {
        	m >>= 1;
        	n >>= 1;
        	shift++;
        }
        return m << shift;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.rangeBitwiseAnd(5, 7));
    }
}

File: #201_Bitwise_AND_of_Numbers_Range/Solution.py
'''
Result of bitwise and of all numbers in the range [m, n] is the common part of their left bit
So we don't need to check all numbers in the range, we can only check the common part of m and n
'''

class Solution(object):
    def rangeBitwiseAnd(self, m, n):
    	if m == 0:
    		return 0
    	count = 0
    	while m != n:
    		m >>= 1
    		n >>= 1
    		count += 1
    	return m << count

sol = Solution()
print sol.rangeBitwiseAnd(3, 5)